engine/src/main/java/org/terasology/rendering/opengl/ScreenGrabber.java

```diff
@@ -112,36 +117,70 @@ public void saveScreenshot() {
-            Path path = PathManager.getInstance().getScreenshotPath().resolve(fileName);
-            BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
-
-            for (int x = 0; x < width; x++) {
-                for (int y = 0; y < height; y++) {
-                    int i = (x + width * y) * 4;
-                    int r = buffer.get(i) & 0xFF;
-                    int g = buffer.get(i + 1) & 0xFF;
-                    int b = buffer.get(i + 2) & 0xFF;
-                    image.setRGB(x, height - (y + 1), (0xFF << 24) | (r << 16) | (g << 8) | b);
-                }
-            }
-
-            try (OutputStream out = new BufferedOutputStream(Files.newOutputStream(path))) {
-                ImageIO.write(image, format, out);
-                logger.info("Screenshot '" + fileName + "' saved! ");
-            } catch (IOException e) {
-                logger.warn("Failed to save screenshot!", e);
-            }
-        };
+        Runnable task;
+        if (savingGamePreview) {
+            task = () -> saveGamePreviewTask(buffer, width, height);
+            this.savingGamePreview = false;
+        } else {
+            task = () -> saveScreenshotTask(buffer, width, height);
+        }
 
         threadManager.submitTask("Write screenshot", task);
         isTakingScreenshot = false;
     }
 
+    /**
+     * Saves given screenshot data to file.
+     *
+     * The file is then saved in the designated screenshot folder with a filename in the form:
+     *
+     *     Terasology-[yyMMddHHmmss]-[width]x[height].[format]
+     */
+    private void saveScreenshotTask(ByteBuffer buffer, int width, int height) {
+        final String format = renderingConfig.getScreenshotFormat();
+        Path screenshotPath = getScreenshotPath(width, height, format);
+        BufferedImage image = getImageDataFromBuffer(buffer, width, height);
+        writeImageToFile(image, screenshotPath, format);
+    }
+
+    /**
+     *  Saves given image data to file in save game folder.
+     */
+    private void saveGamePreviewTask(ByteBuffer buffer, int width, int height) {
+        BufferedImage image = getImageDataFromBuffer(buffer, width, height);
+        writeImageToFile(image, savedGamePath, "jpg");
+    }
+
+    private void writeImageToFile(BufferedImage image, Path path, String format) {
+        try (OutputStream out = new BufferedOutputStream(Files.newOutputStream(path))) {
+            ImageIO.write(image, format, out);
+            logger.info("Screenshot saved to {}! ", path);
+        } catch (IOException e) {
+            logger.warn("Failed to save screenshot!", e);
+        }
+    }
+
+    private BufferedImage getImageDataFromBuffer(ByteBuffer buffer, int width, int height) {
+        BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
+
+        for (int x = 0; x < width; x++) {
```

Comment:
This bit of code needs some comments (at least at the level of "converts ByteBuffer to x by y process")